[AMF] Registration status update

matejGradisar opened the pull request.

matejGradisar commented:
Two traces:

[AMF_RegistrationStatusUpdate.zip](https://github.com/open5gs/open5gs/files/15293278/AMF_RegistrationStatusUpdate.zip)

acetcom commented:
@matejGradisar 

I don't understand the code below, if you have a reason for writing it this way, please elaborate in the comments.

- src/amf/context.c
```
3043     memcpy(amf_ue->old_amf.m_tmsi,
3044             &mobile_identity_guti->m_tmsi, sizeof(uint32_t));
3045     amf_ue->old_amf.guti.m_tmsi = be32toh(mobile_identity_guti->m_tmsi);
```

amf_ue_create_to_release_session_list()/amf_ue_clear_to_release_session_list() is not implemented clearly, which can cause to_release_session_list to be allocated twice.

- src/amf/context.h
```
293     OpenAPI_list_t *to_release_session_list;
...
984 void amf_ue_create_to_release_session_list(amf_ue_t *amf_ue);
985 void amf_ue_clear_to_release_session_list(amf_ue_t *amf_ue);
```

A double allocation of sess->sm_context.ref is possible

- src/amf/gmm-handler.c
```
1681         sess->sm_context.ref = ogs_strdup(PduSessionContext->sm_context_ref
```

The memory for target_guami is unallocated and implemented. How were you able to create the PCAP below? Please double check that this is what you tested with this code.

- src/amf/gmm-sm.c
```
1279                     memcpy(discovery_option->target_guami,
1280                             amf_ue->guami, sizeof(ogs_guami_t));
```

It seems that your code is largely insecure, so providing PCAP alone is not enough. Please provide a test method or provide test code.

Thanks a lot!
Sukchan

matejGradisar commented:
I will check your comments and make corrections.

Regarding your comment "_The memory for target_guami is unallocated and implemented._" - you are right, thank you.

I used a code snippet to trigger "UE contex transfer", a simulation as if the UE's GUTI has the guami of AMF2. A similar simulation was used for "target_guami" in the "discovery_option" for "Registration status update". Based on that simulation I could implement the scenario for "Registration status update" and made the PCAP. When I deleted the simulation and cleaned the code, I did not notice that the "discovery_option" was unallocated.

I had to use the simulation at those points in code (as I wrote in the comments for UE context transfer), because our testing tool does not support the "UE context transfer" scenario.

I will correct the error, it requires minor changes. I cannot promise the tests however, they will take more time.
